zbuf/merger.go

```diff
@@ -0,0 +1,192 @@
+package zbuf
+
+import (
+	"context"
+	"sync"
+
+	"github.com/brimsec/zq/expr"
+	"github.com/brimsec/zq/field"
+	"github.com/brimsec/zq/zng"
+)
+
+// A Merger merges multiple upstream Pullers into one downstream Puller.
+// If the input streams are ordered according to the configured comparison,
+// the output of Merger will have the same order.  Each parent puller is run
+// in its own goroutine so that deadlock is avoided when the upstream pullers
+// would otherwise block waiting for an adjacent puller to finish but the
+// Merger is waiting on the upstream puller.
+type Merger struct {
+	ctx      context.Context
+	cmp      expr.CompareFn
+	cancelCh chan struct{}
+	once     sync.Once
+	pullers  []*puller
+}
+
+type puller struct {
+	Puller
+	ch    chan batch
+	recs  []*zng.Record
+	batch Batch
+}
+
+type batch struct {
+	Batch
+	err error
+}
+
+func NewCompareFn(mergeField field.Static, reversed bool) expr.CompareFn {
+	fn := expr.NewCompareFn(true, expr.NewDotExpr(mergeField))
+	if !reversed {
+		return fn
+	}
+	return func(a, b *zng.Record) int { return fn(b, a) }
+}
+
+func NewMerger(ctx context.Context, pullers []Puller, cmp expr.CompareFn) *Merger {
+	m := &Merger{
+		ctx:      ctx,
+		cmp:      cmp,
+		cancelCh: make(chan struct{}),
+	}
+	m.pullers = make([]*puller, 0, len(pullers))
+	for _, p := range pullers {
+		m.pullers = append(m.pullers, &puller{Puller: p, ch: make(chan batch)})
+	}
+	return m
+}
+
+func (m *Merger) Cancel() {
+	if m.cancelCh != nil {
+		close(m.cancelCh)
+		m.cancelCh = nil
+	}
```

Comment:
Feels like this should be safe for concurrent access. Why not replace the cancelCh with a context.CancelFunc derived from the passed context? Ala on line 47

```golang
func NewMerger(ctx context.Context, pullers []Puller, cmp expr.CompareFn) *Merger {
        ctx, cancel := context.WithCancel(ctx)
 	m := &Merger{
 		ctx:      ctx,
 		cmp:      cmp,
 		cancel: cancel,
 	}
```

Reply:
Either above or do this:
```suggestion
	select {
	case <-m.cancelCh:
	default:
		close(m.cancelCh)
	}
```